File: sdk/samples/WFSamples/Technologies/TransactionalService/TransactionalServiceExample/CS/SampleWorkflow/BalanceTransferWorkflow.cs
//---------------------------------------------------------------------
//  This file is part of the Windows Workflow Foundation SDK Code Samples.
// 
//  Copyright (C) Microsoft Corporation.  All rights reserved.
// 
//This source code is intended only as a supplement to Microsoft
//Development Tools and/or on-line documentation.  See these other
//materials for detailed information regarding Microsoft code samples.
// 
//THIS CODE AND INFORMATION ARE PROVIDED AS IS WITHOUT WARRANTY OF ANY
//KIND, EITHER EXPRESSED OR IMPLIED, INCLUDING BUT NOT LIMITED TO THE
//IMPLIED WARRANTIES OF MERCHANTABILITY AND/OR FITNESS FOR A
//PARTICULAR PURPOSE.
//---------------------------------------------------------------------
using System;
using System.Workflow.Activities;

using System.Collections.Generic;

namespace Microsoft.Samples.Workflow.TransactionalServiceSample
{
    public sealed partial class BalanceTransferWorkflow : SequentialWorkflowActivity
    {
        private int transferAmountValue;

        public int TransferAmount
        {
            get { return transferAmountValue; }
            set { transferAmountValue = value; }
        }

        private IList<Int32> accountBalancesValue;

        public IList<Int32> AccountBalances
        {
            get { return accountBalancesValue; }
            set { accountBalancesValue = value; }
        }

        private int accountNumberValue = 1;

        public int AccountNumber
        {
            get { return accountNumberValue; }
            set { accountNumberValue = value; }
        }

        private Exception workflowExceptionValue = new Exception();

        public Exception WorkflowException
        {
            get { return workflowExceptionValue; }
            set { workflowExceptionValue = value; }
        }
	

        public BalanceTransferWorkflow()
        {
            AccountBalances = new Int32[] { 0, 0 };
            this.CanModifyActivities = true;
            InitializeComponent();
            this.CanModifyActivities = false;
        }
        private void onBefore_CreditAmount(object sender, EventArgs e)
        {
            Console.WriteLine("Crediting the Checking amount in account number {0}",AccountNumber);
            this.creditAmount1.Amount = TransferAmount;
        }
        private void onBefore_DebitAmount(object sender, EventArgs e)
        {
             Console.WriteLine("Debiting the Savings amount in account number {0}",AccountNumber);
             this.debitAmount1.Amount = TransferAmount;
        }
        private void onBeforeTransactionInquiry(object sender, EventArgs e)
        {
             Console.WriteLine("Querying account number {0} to find the balance before the transaction",AccountNumber);
        }

        private void onAfterTransactionInquiry(object sender, EventArgs e)
        {
             Console.WriteLine("Querying account number {0} to find the balance after the transaction",AccountNumber);
        }

        private void onDisplayAccountBalances(object sender, EventArgs e)
        {
            Console.WriteLine("The account balances for account number {0} are:  Checking : {1:c} , Savings : {2:c}",
                 AccountNumber, AccountBalances[0], AccountBalances[1]);
        }

        private void onCodeNotify_ExecuteCode(object sender, EventArgs e)
        {
            WorkflowException = faultHandlerActivity1.Fault;
             Console.WriteLine("The transaction failed due to an exception aborting the transaction.");
             Console.WriteLine(WorkflowException.Message);
        }
    }

}

File: sdk/samples/WFSamples/Technologies/TransactionalService/TransactionalServiceExample/VB/SampleWorkflow/BalanceTransferWorkflow.vb
'---------------------------------------------------------------------
'  This file is part of the Windows Workflow Foundation SDK Code Samples.
' 
'  Copyright (C) Microsoft Corporation.  All rights reserved.
' 
'This source code is intended only as a supplement to Microsoft
'Development Tools and/or on-line documentation.  See these other
'materials for detailed information regarding Microsoft code samples.
' 
'THIS CODE AND INFORMATION ARE PROVIDED AS IS WITHOUT WARRANTY OF ANY
'KIND, EITHER EXPRESSED OR IMPLIED, INCLUDING BUT NOT LIMITED TO THE
'IMPLIED WARRANTIES OF MERCHANTABILITY AND/OR FITNESS FOR A
'PARTICULAR PURPOSE.
'---------------------------------------------------------------------

Imports System
Imports System.ComponentModel
Imports System.ComponentModel.Design
Imports System.Collections
Imports System.Collections.Generic
Imports System.Drawing
Imports System.Workflow.ComponentModel.Compiler
Imports System.Workflow.ComponentModel.Serialization
Imports System.Workflow.ComponentModel
Imports System.Workflow.ComponentModel.Design
Imports System.Workflow.Runtime
Imports System.Workflow.Activities
Imports System.Workflow.Activities.Rules

Public Class BalanceTransferWorkflow
    Inherits SequentialWorkflowActivity

    Private accountBalancesValue As IList(Of Int32) = New Int32(1) {0, 0}
    Private accountNumberValue As Int32 = 1
    Private workflowExceptionValue As Exception = New System.Exception

    Public Property AccountBalances() As IList(Of Int32)
        Get
            Return accountBalancesValue
        End Get
        Set(ByVal value As IList(Of Int32))
            accountBalancesValue = value
        End Set
    End Property

    Public Property AccountNumber() As Int32
        Get
            Return accountNumberValue
        End Get
        Set(ByVal value As Int32)
            accountNumberValue = value
        End Set
    End Property

    Public Property WorkflowException() As Exception
        Get
            Return workflowExceptionValue
        End Get
        Set(ByVal value As Exception)
            workflowExceptionValue = value
        End Set
    End Property

    Private Sub onBeforeTransactionInquiry(ByVal sender As System.Object, ByVal e As System.EventArgs)
        Console.WriteLine("Inquiring account number {0} to find the balance before the transaction", AccountNumber)
    End Sub

    Private Sub onDisplayAccountBalances(ByVal sender As System.Object, ByVal e As System.EventArgs)
        Console.WriteLine("The Account Balances for account number {0} are Checking : {1:c} , Savings {2:c}", AccountNumber, AccountBalances(0), AccountBalances(1))
    End Sub

    Private Sub onBeforeCreditAmount(ByVal sender As System.Object, ByVal e As System.EventArgs)
        Console.WriteLine("Crediting the Checking amount in account number {0}", AccountNumber)
        'Me.TransferAmount = CType((Me.Parameters("TransferAmount").Value), Int32)
        Me.CreditAmount1.Amount = TransferAmount
    End Sub

    Private Sub onBeforeDebitAmount(ByVal sender As System.Object, ByVal e As System.EventArgs)
        Console.WriteLine("Debiting the Savings amount in account number {0}", AccountNumber)
        'Me.TransferAmount = CType((Me.Parameters("TransferAmount").Value), Int32)
        Me.DebitAmount1.Amount = TransferAmount
    End Sub

    Private Sub onAfterTransactionInquiry(ByVal sender As System.Object, ByVal e As System.EventArgs)
        Console.WriteLine("Inquiring account number {0} to find the balance after the transaction", AccountNumber)
    End Sub

    Private Sub onExecuteCode(ByVal sender As System.Object, ByVal e As System.EventArgs)
        WorkflowException = exceptionHandler1.Fault
        Console.WriteLine("The transaction failed due to an exception.  Aborting the transaction ....")
        Console.WriteLine(WorkflowException.Message)
    End Sub

    Private transferAmountValue As Integer
    Property TransferAmount() As Integer
        Get
            Return transferAmountValue
        End Get
        Set(ByVal value As Integer)
            transferAmountValue = value
        End Set
    End Property
End Class
